### ISSUE TITLE
Assertion error in Safari, unless the Web Inspector is open

### ISSUE THREAD

**Issue by WillemSleegers** [r-wasm/webr#189]
I'm having a weird bug that WebR produces an assertion error in Safari but not if I open the Web Inspector. If I have the inspector open, refresh the page, it all works fine. It also works fine in Chrome. Any clues as to what may cause an error like this?

A screenshot of the error message:
<img width="1584" alt="Screenshot 2023-04-04 at 23 29 13" src="https://user-images.githubusercontent.com/4398042/229926989-e286f626-ba23-4e02-9415-9bf7bc160306.png">

A link to the hosted page: 
https://642c95d6d914dd11c91843b1--webr-template.netlify.app

Link to the code:
https://github.com/WillemSleegers/webr-template

**Comment by georgestagg** [comment]
Interesting! No, I have not seen that error before. I think I will need to reproduce so as to learn more.

Your demo website seems to be working for me OK now, running in Safari 16.3 and Safari Tech Preview 16.4. Did you change something? If not, which version of Safari are you currently running?

**Comment by hrbrmstr** [comment]
In Version 16.3 (18614.4.6.1.6) on Ventura (latest stable) I too do not see
an assertion error.

**Comment by WillemSleegers** [comment]
Odd! It's working fine for me now as well. Feel free to close the issue and I'll be back in case it pops up again.

**Comment by WillemSleegers** [comment]
Coincidentally, I just started it up and I have the problem again. Same Netlify link. 

Running Safari Version 16.4 (18615.1.26.11.23 and Ventura 13.3.1 (22E261).

**Comment by georgestagg** [comment]
Still no assertion error here, but I've reopened so we can track this.

**Comment by WillemSleegers** [comment]
Thanks. 

I just tried it on the most recent version of Safari Technology Preview, but same issue. No issue on my iPad though.

**Comment by WillemSleegers** [comment]
I have since taken down the website where I had the issue. I've been messing around with another example and I ran into the same issue.

Hosted page: https://main--unrivaled-cranachan-f0332f.netlify.app
GitHub rep: https://github.com/WillemSleegers/webr-nextjs-example

**Comment by WillemSleegers** [comment]
I should perhaps mention that I also get the following error message:
<img width="1154" alt="image" src="https://github.com/r-wasm/webr/assets/4398042/449b2f3f-84ab-4eeb-b0c9-4a36e582847d">

Despite the error message, the code runs after opening the Web Inspector and in Chrome I have no issues, so it appears to be something Safari specific.

**Comment by hypebright** [comment]
@WillemSleegers might be worth another try. I had the same issue on Safari 16.5.2 (18615.2.9.11.10), but since today I don't get the error anymore

**Comment by georgestagg** [comment]
I have a hunch that this is related to CloudFront caching of webr.r-wasm.org and mismatched versions of the webR TypeScript library and the R.bin.wasm WebAssembly binary being used together, but I have no real evidence.

That would explain the transient nature, if it is the case then the problems would go away once CloudFront's caches are all back in sync with the backing storage.

I guess the right thing to do would be to tweak our CI scripts to force a cache invalidation on webr.r-wasm.org/latest whenever a new commit is pushed to `main` and deployed. This would be something useful to have in any case.

**Comment by baogorek** [comment]
Any updates on this? I have a single HTML file that runs fine in Chrome and Edge, but I'm getting messages similar to @WillemSleegers in Safari. 
![image](https://github.com/user-attachments/assets/9fc54b60-6f9d-4581-af72-545bdca005e6)

My webR code starts like this:

```
  <script type="module">
    import('https://webr.r-wasm.org/latest/webr.mjs').then(async ({ WebR }) => {
      const webR = new WebR();
      await webR.init();
```

**Comment by WillemSleegers** [comment]
Yeah I also still get those messages. Would be nice to resolve that.

**Comment by georgestagg** [comment]
The `Not allowed [...] /web-worker.js.map` and `Cannot load [...] /web-worker.js.map` messages are related to the browser trying to load debugging information (the `.map` file) from a source that has been loaded in a particular way to avoid Cross-Origin restrictions. They should be harmless, though I appreciate are an annoyance.

---

@baogorek Does your code run any computation with webR after `init()`? Even with those messages in the console log, it should still work. Can you share a full self-contained example of a failing script?

**Comment by baogorek** [comment]
@georgestagg , here's an example: https://github.com/baogorek/Miscellaneous/blob/master/R-code/hello-webR.html

I just reallized it halfway works on Safari. On Chrome, you'll see the attached image, although there is a few seconds delay to get the plot. On Safari, you'll see the "Hello, World! x= 12" from the first R code block, but the second part of the block - the plotting portion - failed.
![image](https://github.com/user-attachments/assets/beb9096f-72f1-4878-b512-f149cca4d896)

So it looks like the problem is related to displaying the image:

```
      const rCode2 = `
          webr::install("ggplot2")
          library(ggplot2)
          webr::canvas()
          x <- rnorm(100)
          y <- rnorm(100)
          df <- data.frame(x=x, y=y)
          # Plot with axes, points, and labels all in one frame
          p <- ggplot(df, aes(x=x, y=y)) + geom_point()
          print(p)
          dev.off()
      `;

      await webR.evalRVoid(rCode2);

        for (;;) {
          const output = await webR.read();
          switch (output.type) {
            case 'canvas':
              if (output.data.event === 'canvasImage') {
                canvas.getContext('2d').drawImage(output.data.image, 0, 0);
              } else if (output.data.event === 'canvasNewPage') {
                 console.log("Not doing anything");
              }
              break;
            default:
              console.log(output);
          }
        }
   ```
   
Sorry that it looks like those messages were a red herring, but I guess it did make debugging more challenging. Let me know if you want me to start another issue.

**Comment by georgestagg** [comment]
> Sorry that it looks like those messages were a red herring, but I guess it did make debugging more challenging.

No worries, I understand.

> Let me know if you want me to start another issue.

Yes, that is probably a good idea. When you open the issue please add what version of Safari you see the problems with (Safari -> About Safari in the top menu will tell you). Older versions of Safari didn't support a feature called `OffscreenCanvas`, which we use for plotting. This is potentially the root cause of the problem.

**Comment by WillemSleegers** [comment]
Are there any plans to fix the console errors in Safari? 

I'm also getting the following error message in Chrome: Refused to get unsafe header "Content-Encoding"

**Comment by georgestagg** [comment]
> Are there any plans to fix the console errors in Safari?

Are you still seeing the "unknown taskId 0" error? I'd like to fix the underlying issue but I was never able to reproduce it at my end. WebR currently seems to work OK in Safari for most users.

Do you still have an example deployment causing the issue at your end? Does the error also appear for other deployments, such as at https://webr.r-wasm.org/latest/ or https://shinylive.io/r/examples/?

Is your device an enterprise managed device? I wonder if there is some security software or network proxy causing the issue.

> I'm also getting the following error message in Chrome: Refused to get unsafe header "Content-Encoding"

I think this is coming from Emscripten ([as part of the libfs code](https://github.com/emscripten-core/emscripten/blob/c05d5fd3b488bf67e7f5f4500820ceb5922189c6/src/lib/libfs.js#L1752)). It should not cause any problems, it just means that VFS images will be downloaded in entirety when mounted to the Emscripten virtual filesystem, which is OK.

My understanding is that the "unsafe header" error should go away when the web server hosting webR is configured to send [additional CORS headers](https://docs.r-wasm.org/webr/latest/serving.html).

**Comment by WillemSleegers** [comment]
Sorry for the late response @georgestagg . I have not seen the unknown taskId 0 error in a while, so fingers crossed that's resolved! 

The console errors I was referring to where the ones mentioned by @baogorek (e.g., not allowed to load local resource). I do still see those sometimes, but mainly during local development. Everything seems to work just fine in production.
